docs/adr/0009-infrastructure-as-code-sharing-approach.md

```diff
@@ -0,0 +1,94 @@
+# infrastructure-as-code-sharing-approach
+
+## Status
+
+Proposed
+
+## Context
+
+* The current version of Marain uses ARM templates to automate the infrastructure deployment for each Marain service.
+* The deployment templates resides alongside the code for each service
+* There is significant commonality/duplication of ARM templates between services
+* We wish to migrate to using Bicep to define the Marain infrastructure deployments
+* A desire to minimise duplication and utilise shared Bicep modules
+* Ideally, the overall Marain deployment process should be as low-friction as the current "'git clone' & deploy" approach
+
+Given this context and goals, the challenge is how to combine the use of shared Bicep modules (to de-duplicate the infrastructure-as-code) whilst still having a low friction means of deploying the Marain stack.  The absence of a defacto public Bicep module repository that can host the Marain-specific modules means there isn't an obvious way to share these modules without additional work.
+
+***NOTE**: A more general, but related, Bicep reuse [ADR](https://github.com/endjin/Endjin.RecommendedPractices.Bicep/blob/main/docs/adr/0002-sharing-bicep-modules.md) was written before it became clear that the public Bicep registry would not be 'open to all' in the same way that nuget.org or Docker Hub are.*
+
+In the ideal world the Marain deployment would use Bicep modules in the [Endjin.RecommendedPractices.Bicep repository](https://github.com/endjin/Endjin.RecommendedPractices.Bicep) so as to avoid duplication as well as a set of Marain-specific modules that are re-usable across different Marain services.  This means that the envisioned deployment process would need to interact with 3 sets of Bicep modules:
+
+1. Shared/Common - non-Marain specific functionality 
+1. Marain shared - functionality common to multiple Marain services
+1. Marain service-specific - functionality tailored to an individual Marain service
+
+This ADR considers the following options:
+
+1. Hosting a public Bicep registry
+1. Supporting internal Bicep registries
+1. Shipping pre-compiled ARM templates
+
+## Options
+
+### Hosting a public Bicep registry
+
+This option involves us hosting an Azure Container Registry that we make publicly available and publishing the Marain Bicep modules to it as part of the overall Marain release process.
```

Comment:
Is this an officially supported use of an ACR?

I don't know what Bicep's relationship with a registry looks like. Here are some things I don't know:

* is Bicep designed with the intention that it will be used with a mixture of "the public Bicep registry" referred to earlier and other registries (such as a private one, or a non-Microsoft public one)? E.g., does its configuration fully support an idea of "here are all the registries we're using" or do you end up doing something more hacky like pointing it to a registry which somehow defers back to the public registry (e.g., as you can do with Azure DevOps 'artifact feeds')?
* is it just making simple HTTP requests based on a presumed URL structure? (E.g., could you in principle use a blob store as a registry?) Or are there particular things it requires of a repository?
* what are the expectations around auth when the bicep tools talk to a registry? might these change over time?

In particular, I'm thinking about the fact that when NuGet first launched, it really did treat the repo as just a file store, so you could use pretty much any web server as a package server, but that's not exactly true any more.

So I'm wondering: does an ACR work as a Bicep store just because it happens to resemble a generic static HTTP service enough to keep bicep happy? (Bicep is not Docker. A Bicep module is not a container image. So it's not obvious to me that a container registry should necessarily work as a Bicep registry.) And if so, might Bicep's expectations change in the future in a way that means this won't always be true.